### DISCUSSION TITLE
Add imageOffchainData & nftTitle

### DISCUSSION THREAD

**Discussion by gege83var** [capsule-corp-ternoa/ternoa-subql#171] [Ideas]
Hello guys,

I would like to suggest to add 2 fields in nftEntities.
I don't know if it's possible and easy to add or no (cause ipfs data can be anything)
But, I think adding them can be really useful to query image & or nft title 100 by 100 instead of getting 100 nft offchain data and request 100 time to IPFS there name & image (at least)

That could be drastically decrease the time to show NFT in wallet, marketplace, ... anywhere we need to show NFT finally

More than, (it's already possible, but not perfect) is to request groupBy imageOffchainData to show only 1 time the image if is not unique (Happy Tiime collection for exemple) 
That's work great with offchainData

But if someone create 2 different NFT but with the same offchainData image link it doesn't work (very rare case I know)
But it's another utility of these request

Thanks a lot <3

**Comment by ipapandinas** [comment]
Hello @gege83var

It would be super convenient to improve developer experience, that's a fact.
The problem is that we need to switch our Indexer service in `--unsafe` mode to making external requests to IPFS. As explained [here by the Subquery team](https://academy.subquery.network/run_publish/references.html#unsafe-node-service):

> Unsafe mode controls various features that compromise the determinism of a SubQuery project. It makes it impossible to guarantee that the data within two identical projects running independently, will be 100% consistent.

Here is an idea of flow that I suggest at the moment before we find a better solution:

### Backend flow to get NFT data using an additionnal local DB

1. Parse query from the HTTP request
2. Validate query with an object schema validation (e.g. JOI)
3. Get NFT on-data from Ternoa Indexer
4. Populate NFT metadatas from the local DB
5. If hash not found in the local DB - Get metadatas from IPFS (with an optional retry mechanism) + save metadatas into the local DB

This flow is currently used by [Secret Stash](https://secret-stash.io/).
I don't know how metadatas are handled by the Wallet team; maybe @dhammani can share us additional tips&tricks.

Thank you so much for this nice suggestions ✨
